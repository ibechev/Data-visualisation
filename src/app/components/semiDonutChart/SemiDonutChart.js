import React from "react";
import PropTypes from "prop-types";

const SemiDonutChart = ({ name, score, max }) => {
  const degrees = ((score / max) * 100 * 180) / 100;

  const arcFill = {
    WebkitTransform: `rotate(${degrees}deg)`,
    msTransform: `rotate(${degrees}deg)`,
    transform: `rotate(${degrees}deg)`
  };

  return (
    <div className="semi-donut-chart-container">
      <div className="donut-wrapper">
        <div className="arc" style={arcFill} />
        <span className="score">{score}</span>
      </div>
      <span className="donut-chart-label">{name}</span>
    </div>
  );
};

SemiDonutChart.propTypes = {
  name: PropTypes.string.isRequired,
  score: PropTypes.number.isRequired,
  max(props, propName, component) {
    if (!props.hasOwnProperty(propName))
      return new Error(`Missing prop ${propName} at component ${component}.`);

    if (typeof props[propName] !== "number")
      return new Error(`Prop ${propName} must be of type number.`);

    if (props[propName] < props.score)
      return new Error(`Prop ${propName} can not be less than prop score.`);
  }
};

export default SemiDonutChart;
