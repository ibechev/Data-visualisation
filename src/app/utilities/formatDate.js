export const formatDate = date => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ];
  const tempDate = date.split("-");
  return `${months[parseInt(tempDate[1]) - 1]} ${tempDate[0].substring(2)}`;
};
