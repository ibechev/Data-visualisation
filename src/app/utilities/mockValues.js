export const mockAlerts = [
  {
    name: "Adam Campton",
    event: "end of year review",
    eventDate: "2019-01-18",
    importance: "very"
  },
  {
    name: "Emma O'Sullivan",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Brandon Walsh",
    event: "end of year review",
    eventDate: "2019-02-06",
    importance: "middle"
  },
  {
    name: "David O'Brien",
    event: "end of year review",
    eventDate: "2019-02-23",
    importance: "very"
  },
  {
    name: "Sophie O'Connor",
    event: "end of year review",
    eventDate: "2019-01-03",
    importance: "low"
  },
  {
    name: "Sean Connor",
    event: "end of year review",
    eventDate: "2019-01-05",
    importance: "very"
  },
  {
    name: "Gery Brown",
    event: "end of year review",
    eventDate: "2019-01-12",
    importance: "middle"
  },
  {
    name: "Stephen Greenday",
    event: "end of year review",
    eventDate: "2019-02-11",
    importance: "very"
  },
  {
    name: "Lisa Aniston",
    event: "end of year review",
    eventDate: "2018-12-27",
    importance: "low"
  },
  {
    name: "Gordon Niels",
    event: "end of year review",
    eventDate: "2018-12-28",
    importance: "middle"
  }
];

export const mockCustomerInfo = {
  max: 270,
  data: [
    { month: "Dec-2017", value: 99 },
    { month: "Jan-2018", value: 80 },
    { month: "Feb-2018", value: 71 },
    { month: "Mar-2018", value: 96 },
    { month: "Apr-2018", value: 80 },
    { month: "May-2018", value: 115 },
    { month: "Jun-2018", value: 146 },
    { month: "Jul-2018", value: 121 },
    { month: "Aug-2018", value: 160 },
    { month: "Sep-2018", value: 168 },
    { month: "Oct-2018", value: 185 },
    { month: "Nov-2018", value: 203 }
  ]
};
