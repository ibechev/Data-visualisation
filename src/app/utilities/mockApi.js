import { mockAlerts, mockCustomerInfo } from "../utilities/mockValues";

const randomTimeRange = (min, max) => {
  return (Math.floor(Math.random() * (max - min + 1)) + min) * 1000;
};
const simulateLatency = ms => new Promise(resolve => setTimeout(resolve, ms));

export const mockFetchAlerts = async () => {
  await simulateLatency(randomTimeRange(0, 3));
  return mockAlerts.map(alert => ({ ...alert }));
};

export const mockFetchClientsInfo = async () => {
  const { max, data } = mockCustomerInfo;
  await simulateLatency(randomTimeRange(0, 3));
  return {
    max,
    data: data.map(each => ({ ...each }))
  };
};
