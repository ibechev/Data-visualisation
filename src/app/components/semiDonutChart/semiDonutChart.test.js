import React from "react";
import { shallow } from "enzyme";
import SemiDonutChart from "./SemiDonutChart";
import renderer from "react-test-renderer";
import { mockCustomerInfo } from "../../utilities/mockValues";
import { propTypeErrors } from "../../../../tests/config/testUtils";

describe("Expect <SemiDonutChart />", () => {
  const defaultProps = {
    name: "Semi donut chart",
    score: 110,
    max: 200
  };

  const wrapper = shallow(<SemiDonutChart {...defaultProps} />);

  it("to render correctly", () => {
    const wrapper = renderer
      .create(<SemiDonutChart {...defaultProps} />)
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to not throw error with correct props", () => {
    propTypeErrors(<SemiDonutChart {...defaultProps} />, defaultProps);
  });

  it("to set the correct style of <div class='arc'", () => {
    const degrees = ((defaultProps.score / defaultProps.max) * 100 * 180) / 100;
    expect(wrapper.find("div.arc").prop("style")).toHaveProperty(
      "transform",
      `rotate(${degrees}deg)`
    );
  });

  it("to display prop 'score' in <span class='score'", () => {
    expect(wrapper.find("span.score").text()).toEqual(
      defaultProps.score.toString()
    );
  });

  it("to display prop 'name' in <span class='donut-chart-label' />", () => {
    expect(wrapper.find("span.donut-chart-label").text()).toEqual(
      defaultProps.name
    );
  });
});
