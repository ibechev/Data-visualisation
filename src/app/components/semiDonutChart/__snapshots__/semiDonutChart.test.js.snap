// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <SemiDonutChart /> to render correctly 1`] = `
<div
  className="semi-donut-chart-container"
>
  <div
    className="donut-wrapper"
  >
    <div
      className="arc"
      style={
        Object {
          "WebkitTransform": "rotate(99.00000000000001deg)",
          "msTransform": "rotate(99.00000000000001deg)",
          "transform": "rotate(99.00000000000001deg)",
        }
      }
    />
    <span
      className="score"
    >
      110
    </span>
  </div>
  <span
    className="donut-chart-label"
  >
    Semi donut chart
  </span>
</div>
`;
